WalletWasabi/WabiSabi/Backend/Rounds/Arena.cs

```diff
@@ -254,99 +307,144 @@ private void TimeoutAlices()
+							Config,
+							request.RoundId,
+							alice,
+							request.ZeroAmountCredentialRequests,
+							request.ZeroVsizeCredentialRequests);
+
+						// Now that alice is in the round, make it available by id.
+						(AlicesById as IDictionary<uint256, Alice>).Add(alice.Id, alice);
+
+						return response;
+					}
+					catch
+					{
+						AlicesByOutpoint.Remove(coin.Outpoint, out _);
+						throw;
+					}
+				}
 			}
 		}
 
 		public async Task ReadyToSignAsync(ReadyToSignRequestRequest request, CancellationToken cancellationToken)
 		{
-			using (await AsyncLock.LockAsync(cancellationToken).ConfigureAwait(false))
+			if (!RoundsById.TryGetValue(request.RoundId, out var round))
 			{
-				if (Rounds.FirstOrDefault(r => r.Id == request.RoundId) is not Round round)
-				{
-					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.RoundNotFound, $"Round ({request.RoundId}) not found.");
-				}
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.RoundNotFound, $"Round ({request.RoundId}) not found.");
+			}
 
-				if (round.Alices.FirstOrDefault(a => a.Id == request.AliceId) is not Alice alice)
-				{
-					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.AliceNotFound, $"Round ({request.RoundId}): Alice id ({request.AliceId}).");
-				}
+			if (!AlicesById.TryGetValue(request.AliceId, out var alice) || alice.Round != round)
+			{
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.AliceNotFound, $"Round ({request.RoundId}): Alice ({request.AliceId}) not found.");
+			}
 
+			using (await alice.AsyncLock.LockAsync(cancellationToken).ConfigureAwait(false))
+			{
 				var coinJoinInputCommitmentData = new CoinJoinInputCommitmentData("CoinJoinCoordinatorIdentifier", request.RoundId);
 				if (!OwnershipProof.VerifyCoinJoinInputProof(request.OwnershipProof, alice.Coin.TxOut.ScriptPubKey, coinJoinInputCommitmentData))
 				{
 					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WrongOwnershipProof);
 				}
 
-				alice.ReadyToSign = true;
+				using (await AsyncLock.LockAsync(cancellationToken).ConfigureAwait(false))
+				{
+					alice.ReadyToSign = true;
+				}
 			}
 		}
 
 		public async Task RemoveInputAsync(InputsRemovalRequest request, CancellationToken cancellationToken)
 		{
-			using (await AsyncLock.LockAsync(cancellationToken).ConfigureAwait(false))
+			if (!RoundsById.TryGetValue(request.RoundId, out var round))
 			{
-				if (Rounds.FirstOrDefault(x => x.Id == request.RoundId) is not Round round)
-				{
-					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.RoundNotFound, $"Round ({request.RoundId}) not found.");
-				}
-				if (round.Phase != Phase.InputRegistration)
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.RoundNotFound, $"Round ({request.RoundId}) not found.");
+			}
+
+			if (!AlicesById.TryGetValue(request.AliceId, out var alice) || alice.Round != round)
+			{
+				return;
+			}
```

Comment:
Hmm. It already does, because it is a hash of the ownership proof, and the ownership proof commitment data contains the round ID.

Technically this means the `RoundId` request parameter is redundant for any alice related request, if we removedit then the `alice.Round != round` checks wouldn't be necessary, and we could also avoid a lookup on `RoundsById` and just trust `alice.Round` is correct.

I still want to make `RoundId` into a path parameter for several reasons, and one of them is slightly at odds with removing it entirely, but I think it might still work out nicely... I will add as sub-items to the REST bikeshedding TODO item in the task ticket.

Reply:
yes, exactly, that's my point (it is just that i didn't realize it was already committed).